File: 2130. Maximum Twin Sum of a Linked List.js
var pairSum = function(headNode) {

    // Find the middle of the linked list
    let slowPointer = headNode;
    let fastPointer = headNode;

    while (fastPointer && fastPointer.next) {

        slowPointer = slowPointer.next;

        fastPointer = fastPointer.next.next;
    }

    // Reverse the second half of the list
    let previousNode = null;

    while (slowPointer) {

        let nextNode = slowPointer.next;

        slowPointer.next = previousNode;

        previousNode = slowPointer;

        slowPointer = nextNode;
    }

    // Store the maximum twin sum
    let maximumTwinSum = 0;

    let firstHalfPointer = headNode;

    let secondHalfPointer = previousNode;

    // Compare corresponding nodes
    while (secondHalfPointer) {

        let currentTwinSum =
            firstHalfPointer.val +
            secondHalfPointer.val;

        maximumTwinSum = Math.max(
            maximumTwinSum,
            currentTwinSum
        );

        firstHalfPointer =
            firstHalfPointer.next;

        secondHalfPointer =
            secondHalfPointer.next;
    }

    return maximumTwinSum;
};

File: 2130. Maximum Twin Sum of a Linked List.java
class Solution {

    public int pairSum(ListNode headNode) {

        // Find the middle of the linked list
        ListNode slowPointer = headNode;
        ListNode fastPointer = headNode;

        while (fastPointer != null &&
               fastPointer.next != null) {

            slowPointer = slowPointer.next;

            fastPointer = fastPointer.next.next;
        }

        // Reverse the second half of the linked list
        ListNode previousNode = null;

        while (slowPointer != null) {

            ListNode nextNode =
                    slowPointer.next;

            slowPointer.next =
                    previousNode;

            previousNode =
                    slowPointer;

            slowPointer =
                    nextNode;
        }

        // Store the maximum twin sum
        int maximumTwinSum = 0;

        ListNode firstHalfPointer =
                headNode;

        ListNode secondHalfPointer =
                previousNode;

        // Calculate twin sums
        while (secondHalfPointer != null) {

            int currentTwinSum =
                    firstHalfPointer.val +
                    secondHalfPointer.val;

            maximumTwinSum = Math.max(
                    maximumTwinSum,
                    currentTwinSum
            );

            firstHalfPointer =
                    firstHalfPointer.next;

            secondHalfPointer =
                    secondHalfPointer.next;
        }

        return maximumTwinSum;
    }
}
